frame/support/src/traits/hooks.rs

```diff
@@ -136,13 +136,42 @@ pub trait OnRuntimeUpgrade {
 		Weight::zero()
 	}
 
+	/// Same as [`on_runtime_upgrade`], but perform the optional [`pre_upgrade`] and
+	/// [`post_upgrade`] as well.
+	#[cfg(feature = "try-runtime")]
+	fn try_on_runtime_upgrade(checks: bool) -> Weight {
+		let maybe_state = if checks {
+			let _guard = frame_support::StorageNoopGuard::default();
+			let state = Self::pre_upgrade().unwrap();
+			drop(_guard);
+			Some(state)
+		} else {
+			None
+		};
+
+		let weight = Self::on_runtime_upgrade();
+
+		if checks {
+			let _guard = frame_support::StorageNoopGuard::default();
+			// we want to panic if any checks fail right here right now.
```

Comment:
We could, but I am not a fan of it. I think having a backtrace that tells us exactly this line of code paniced is better than bubbling the error up and needing to deal with it. Ofc, this is only a valid argument because this is a testing CLI.